tools/README.md

```diff
@@ -389,3 +385,90 @@ or directly with `python`:
 $ ./tools/sometools/mytool.py -h
 ...
 ```
+
+### Using the `tools.base.checker.Checker` class
+
+A base class for writing checkers (for example, linting tools) has also been provided.
+
+Any classes subclassing `tools.base.checker.Checker` should provide a tuple of `__class__.checks`.
+
+For each named check in `checks` the `Checker` will expect a method of the same name with the prefix `check_`.
+
+For example, setting `checks` to the tuple `("check1", "check2")` the `Checker` will run the methods `check_check1` and `check_check2` in order.
+
+Let's look at an example.
+
+First, we need to add the bazel target.
+
+Edit `tools/sometools/BUILD` and add a `tools.sometools.mychecker` target with a dependency on the base `Checker`.
+
+```starlark
+envoy_py_binary(
+    name = "tools.sometools.mychecker",
+    deps = [
+        "//tools/base:checker",
+    ],
+)
+```
+
+Next add the `MyChecker` class to `tools/sometools/mychecker.py` as follows:
+
+```python
+#!/usr/bin/env python3
+
+import sys
+
+from tools.base.checker import Checker
+
+
+class MyChecker(Checker):
+    checks = ("check1", "check2")
+
+    def check_check1(self) -> None:
+        # checking code for check1
+        try:
+            do_something()
+        except NotSuchABadError:
+            self.warn("check1", ["Doing something didn't work out quite as expected 8/"])
+        except ATerribleError:
+            self.error("check1", ["Oh noes, something went badly wrong! 8("])
+        else:
+            self.succeed("check1", ["All good 8)"])
+
+    def check_check2(self) -> None:
+        # checking code for check2
+        try:
+            do_something_else()
+        except NotSuchABadError:
+            self.warn("check2", ["Doing something else didn't work out quite as expected 8/"])
+        except ATerribleError:
+            self.error("check2", ["Oh noes, something else went badly wrong! 8("])
+        else:
+            self.succeed("check2", ["All good 8)"])
+
+
+def main(*args) -> int:
+    return MyChecker(*args).run()
+
+
+if __name__ == "__main__":
+    sys.exit(main(*sys.argv[1:]))
+```
+
+Just like with the `Runner` class described [above](#using-the-toolsbaserunnerrunner-class) you can
+user both with and without `bazel`. To use without, you will need make it executable, and the end
+user will need to have any dependencies locally installed.
+
+Notice in the check methods the results of the check are logged to one of `self.error`, `self.warn`,
+or `self.succeed`. Each takes a `list` of outcomes. The results will be summarized to the user at the
+end of all checks.
+
+Just like with `Runner` a help menu is automatically created, and you can add custom arguments if
+required.
+
+Also like `Runner`, any added `Checker` classes are expected to have unit tests, and a `pytest_mychecker` bazel target
```

Comment:
Are we going to migrate https://github.com/envoyproxy/envoy/tree/main/tools/code_format over to this one day? Even if not, can we have new checkers written in the framework here rather than that?

Reply:
its certainly my hope that we standardize the tooling using the code in this pr, and shift tests to pytest. Its pretty doable

Reply:
TBH, I think this would be one of the biggest wins; the code formatting tools have hit scaling limits and are hard to extend or reason about at this point. The existing structure has served us well but we should move to something based on your framework going forward.